design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,506 @@
+..
+   Copyright (c) 2023, Arm Limited and its affiliates.  All rights reserved.
+   CC-BY-SA-4.0 AND Apache-Patent-License
+   See LICENSE file for details
+
+Rationale Document for ABI related to the C23 _BitInt type.
+***********************************************************
+
+Preamble
+========
+
+Background
+----------
+
+This document describes the rationale behind the ABI choices made for using the
+bit-precise integral types defined in C2x.  These are ``_BitInt(N)`` and
+``unsigned _BitInt(N)``.  These are defined for integral ``N`` and each ``N`` is
+a different type.
+
+The proposal for these types can be found in following link.
+https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2763.pdf
+
+As the rationale in that proposal mentioned, some applications have uses for a
+specific bit-width type.  In the case of writing C code which can be used to
+determine FPGA hardware these specific bit-width types can lead to large
+performance and space savings.
+
+From the perspective of the Arm ABI we have some trade-offs and decisions to
+make:
+
+- We need to choose a representation for these objects in registers.
+- We need to choose a representation, size and alignment of these objects in memory.
+
+The main trade-offs we have identified in this case are:
+
+- Performance of different C-level operations.
+- Whether certain hardware-level atomic operations are possible.
+- Size cost of storing values in memory.
+- General familiarity of programmers with the representation.
+
+Since this is a new type there is large uncertainty on how it will be used by
+programmers in the future.  Decisions we make here may also influence future
+usage.  Nonetheless we must make trade-off decisions with this uncertainty.  The
+below attempts to analyze possible use-cases to make our best guess as to how
+these types may be used when targeting Arm CPU's.
+
+
+Use-cases known of so far
+-------------------------
+
+There seem to be two different regimes for these types.  The "small" regime
```

Comment:
```suggestion
There are two different regimes for these types.  The "small" regime
```

Reply:
I adjusted to "We believe there are two different regimes" (point being that we could be mistaken and want to convey that this is how we view things rather than exactly what is happening).